### ISSUE TITLE
Mozart vs PHP Scoper, and Autoloader behavior question

### ISSUE THREAD

**Issue by Luc45** [coenjacobs/mozart#94]
Howdy!

I need to prefix the Composer dependencies of a WordPress plugin, I've been studying the differences between Mozart and PHP Scoper and trying to find out what's best for me.

I see that WooCommerce [will be using](https://github.com/woocommerce/woocommerce/pull/28147) Mozart on their next releases, while [Yoast](https://developer.yoast.com/blog/safely-using-php-dependencies-in-the-wordpress-ecosystem/) and [Google Site Kit](https://github.com/google/site-kit-wp/pull/696) are using PHP Scoper.

PHP Scoper is a tool focused on Phars, therefore it doesn't seem to be able to handle Composer Autoloader gracefully.

**Yoast implementation of PHP-Scoper**
Prefixes the `vendor` libraries, remove them from `composer.json` and run `composer dump-autoload`. Then it implements their own `spl_autoload_register` to autoload the prefixed libraries.

**Google Site Kit implementation of PHP-Scoper**
Seems to drop Composer Autoloader entirely, implementing their own `spl_autoload_register`.

I don't like to add `spl_autoload_register` because it adds a file read for every class lookup, and disks are the slowest thing in computing.

I understand that Mozart is designed to work alongside Composer, using the Composer Autoloader itself. Is that correct?

Would you recommend Mozart or PHP Scoper for prefixing a WordPress plugin? Why? I saw that you (Coen Jacobs) had written an article about this subject on your blog, but the link is dead.

Thanks for your time, for this library, and for sharing your knowledge.

**Comment by Luc45** [comment]
Found the blog article with the dead link: https://coenjacobs.me/2019/07/05/why-i-felt-there-was-a-need-for-mozart/

**Comment by coenjacobs** [comment]
Obviously, I prefer Mozart for WordPress projects. I know about the other implementations and in #96 (including some more information, which was once published on my now dead blog) I have explained some of the philosophy and choices behind Mozart. Fun fact, PHP-Scoper didn't even exist back when I started Mozart. That obviously doesn't mean that you should or shouldn't use it, but that's one of the main reasons Mozart now exists.

On the topic of autoloading, yes I do believe that Mozart has a unique approach here. Mozart relies on doing as _little as possible_. In practice, this means that Mozart doesn't change the way Composer is autoloading anything and thus the default autoloader, provided by Composer is used in your project, after Mozart is done transforming the files.
